options.go

```diff
@@ -262,19 +279,123 @@ func setupUnixConn(u *url.URL) (*Options, error) {
 		return nil, errors.New("redis: empty unix socket path")
 	}
 	o.Addr = u.Path
-
 	o.Username, o.Password = getUserPassword(u)
+	return setupConnParams(u, o)
+}
 
-	dbStr := u.Query().Get("db")
-	if dbStr == "" {
-		return o, nil // if database is not set, connect to 0 db.
+type queryOptions struct {
+	q   url.Values
+	err error
+}
+
+func (o *queryOptions) string(name string) string {
+	vs := o.q[name]
+	if len(vs) == 0 {
+		return ""
 	}
+	delete(o.q, name) // enable detection of unknown parameters
+	return vs[len(vs)-1]
+}
 
-	db, err := strconv.Atoi(dbStr)
-	if err != nil {
-		return nil, fmt.Errorf("redis: invalid database number: %w", err)
+func (o *queryOptions) int(name string) int {
+	s := o.string(name)
+	if s == "" {
+		return 0
+	}
+	i, err := strconv.Atoi(s)
+	if err == nil {
+		return i
+	}
+	if o.err == nil {
+		o.err = fmt.Errorf("redis: invalid %s number: %s", name, err)
+	}
+	return 0
+}
+
+func (o *queryOptions) duration(name string) time.Duration {
+	s := o.string(name)
+	if s == "" {
+		return 0
+	}
+	// try plain number first
+	i, err := strconv.Atoi(s)
+	if err == nil {
+		if i == -1 {
+			// special case: -1 is used to disable timeouts
+			return time.Duration(i)
```

Comment:
I am not sure about this one. We need `-1` in Go because `0` is also a default value. But for DSN I think it should be:
- `` - default timeout
- `?read_timeout=0` - disable timeout
- `?read_timeout=1` - 1 second

What do you think?

Reply:
Sure, I can change this, but this would create a difference in configuring the connection:

| direct assignment | URL parameter | behaviour (after `init()`) |
|:-----|:-----|:-----|
| `opt.ReadTimeout = -1` | `?read_timeout=0` | timeout disabled |
| `opt.ReadTimeout = 0` | absent | default timeout |
| `opt.ReadTimeout = 3` | `?read_timeout=3` | 3s timeout |

Personally I don't like that the value depends on the context. I feel this is more consistent:

| direct assignment | URL parameter | behaviour (after `init()`) |
|:-----|:-----|:-----|
| `opt.ReadTimeout = -1` | `?read_timeout=-1` | timeout disabled |
| `opt.ReadTimeout = 0` | `?read_timeout=0` or absent  | default timeout |
| `opt.ReadTimeout = 3` | `?read_timeout=3` | 3s timeout |

Thinking about it though, `?read_timeout=0` implying a default timeout feels wrong too...

Reply:
Yeah, that sounds reasonable.

Reply:
Should negative timeout values be discouraged at all (i.e. yield an error), or should they treated as "disable timeout"?

Reply:
Let's treat as disable timeout to mimic `Options` behavior.